docs/proposals/PROPOSAL_A.md

```diff
@@ -0,0 +1,289 @@
+manifest does not exist in the repository. The list MUST be empty
+if there are no artifacts referencing the given manifest.
+
+**example result of artifacts that reference the `hello-world@sha256:abc123` image:**
+
+```jsonc
+{
+  "referrers": [
+    {
+      "digest": "sha256:aaabbb",
+      "mediaType": "application/vnd.oci.artifact.manifest.v1+json",
+      "artifactType": "signature/example",
+      "size": 312
+    },
+    ...
+  ]
+}
+```
+
+**example result for a manifest that has no artifacts referencing it:**
+
+```json
+{
+  "referrers": []
+}
+```
+
+### Paging Results
+
+The `referrers` API MUST provide for paging, returning a list of [artifact descriptors](./descriptor.md).
+Page size can be specified by adding a `n` parameter to the request URL, indicating that the response should be limited to `n` results.
+
+- If specified, servers MAY return up to `n` items from the entire result set.
+- When `n` is not provided, servers MAY return a default number of items, which may be implementation specific.
+
+A paginated flow begins as:
+
+**template:**
+
+```rest
+GET /v2/{repository}/_oci/artifacts/referrers?digest={digest}&n=<integer>
+```
+
+**expanded example:**
+
+```rest
+GET /v2/{repository}/_oci/artifacts/referrers?digest=sha256:abc123&n=10
+```
+
+The above specifies that a referrers response should be returned limiting the number of results to `n`.
+The response to such a request would look as follows:
+
+```jsonc
+200 OK
+RefType-Api-Version:reftype/1.0
+Link: <url>; rel="next"
+{
+  "referrers": [
+    {
+      "digest": "<string>",
+      "mediaType": "<string>",
+      "artifactType": "<string>",
+      "size": <integer>
+    },
+    ...
+  ]
+}
+```
+
+The above includes up to `n` items from the result set. If there are more items, the URL for the next collection is
+encoded in a [RFC5988][rfc5988] `Link` header, as a "next" relation. Clients SHOULD treat this as an opaque value and not try to
+construct it themselves.
+
+- The presence of the `Link` header communicates to the client that the server has more items. Clients are expected
+  to follow the link to fetch the next page of items, irrespective of the number of items received in the current
+  response.
+- If the header is not present, clients can assume that all items have been received.
+
+> NOTE: In the request template above, the brackets around the url are required.
+
+For example, if the url is:
+
+```
+http://example.com/v2/hello-world/_oci/artifacts/referrers?digest=sha256:abc123&n=5&nextToken=abc
+```
+
+The value of the header would be:
+
+```
+<http://example.com/v2/hello-world/_oci/artifacts/referrers?digest=sha256:abc123&n=5&nextToken=abc>; rel="next"`.
+```
+
+Please see [RFC5988][rfc5988] for details.
+
+### Sorting Results
+
+The `/referrers` API MUST allow for artifacts to be sorted by the date and time in which they were created, which SHOULD be included in the artifact manifest's list of `annotations`.
```

Comment:
Reading this sentence, I'm confused about who SHOULD be including the created annotation; is it the registry? I would assume clients SHOULD set this, but also it might not make sense for certain types of things that get added.

I think making sorting and paging SHOULDs (or even MAYs) would lower the barrier to adoption for both clients and registries.

Reply:
The referrers API MUST support sorting to provide a reasonable set of consistency across registries in a promotion workflow. To your point, not all artifacts will have the `created` annotation. I don't know why they wouldn't but it's not forcing this on the artifact authors. The spec goes on to clarify the behavior for manifests that don't provide the value.

> The artifact's creation time MUST be the value of the `org.oci.artifact.created` annotation.
> The results of the `/referrers` API MUST list artifacts that were created more recently first.
> Artifacts that do not have the `org.oci.artifact.created` annotation MUST appear after those with creation times specified in the list of results.
> There is no specified ordering for artifacts that do not include the creation time in their list of `annotations`.